public/javascripts/AchievementTracker.js

```diff
@@ -0,0 +1,214 @@
+        // Find next badge to unlock.
+        let mapLevelsToBadge = this.mapBadges[badgeType]
+        let nextBadgeToUnlock = null;
+        if (badgeType in this.mapBadges) {
+            let mapLevelsToBadge = this.mapBadges[badgeType]
+
+            for (let level of Object.keys(mapLevelsToBadge)) {
+                let badge = mapLevelsToBadge[level];
+                if (badge.threshold > curValue) {
+                    nextBadgeToUnlock = badge;
+                    break;
+                }
+            }
+        }
+
+        let htmlStatement = '';
+
+        if (nextBadgeToUnlock != null) {
+            let diffValue = nextBadgeToUnlock.threshold - curValue;
+            let curBadge = null;
+            const curBadgeLevel = nextBadgeToUnlock.level - 1;
+            let fractionComplete = 0;
+
+            // Figure out how close they are to the next badge.
+            if (curBadgeLevel in mapLevelsToBadge) {
+                curBadge = mapLevelsToBadge[curBadgeLevel];
+                const badgeValueRange = nextBadgeToUnlock.threshold - curBadge.threshold;
+                fractionComplete = (curValue - curBadge.threshold) / badgeValueRange;
+            } else {
+                fractionComplete = curValue / nextBadgeToUnlock.threshold;
+            }
+
+            // Add an encouraging statement based on how close they are to the next badge level.
+            if (fractionComplete > 0.95) {
+                htmlStatement += i18next.t('dashboard:so-close') + ' ' + i18next.t('dashboard:just') + ' ';
+            } else if (fractionComplete > 0.85) {
+                htmlStatement += i18next.t('dashboard:wow-almost-there') + ' ' + i18next.t('dashboard:just') + ' ';
+            } else if (fractionComplete > 0.1 || curBadgeLevel > 0) {
+                let randStatement = encouragingStatements[Math.floor(Math.random() * encouragingStatements.length)];
+                htmlStatement += i18next.t(randStatement) + ' ';
+            }
+
+            // Convert to from miles to kilometers if using metric system.
+            const measurementSystem = i18next.t('common:measurement-system');
+            if (badgeType === BadgeTypes.Distance && measurementSystem === 'metric') diffValue *= 1.60934;
+
+            // Get the appropriate distance unit, e.g., mission/misión, missions/misiones, labels/etiquetas.
+            let unitTranslation;
+            if (diffValue === 1) unitTranslation = 'dashboard:badge-' + badgeType + '-singular';
+            else unitTranslation = 'dashboard:badge-' + badgeType + '-plural';
+
+            const firstOrNextTranslation = curBadgeLevel === 0 ? 'dashboard:first' : 'dashboard:next';
+
+            // Add translation for how much is left before the next achievement. For example, "1 misión más hasta tu
+            // próximo logro." or "1.3 more miles until your next achievement."
+            htmlStatement += i18next.t('dashboard:more-unit-until-achievement', {
```

Comment:
I reorganized this a bit. Instead of building each piece of the sentence up from translated fragments, I put the entire sentence into one translation, and the translation requires some parameters which do what you were doing.